torch/distributed/_tools/fsdp2_memory_tracker.py

```diff
@@ -0,0 +1,539 @@
+import math
+from collections import defaultdict
+from dataclasses import dataclass
+from datetime import timedelta
+from enum import Enum
+from functools import wraps
+from typing import Any, Callable, Dict, List, NamedTuple, Optional, Union
+
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+from torch._C._distributed_c10d import Work
+from torch._guards import active_fake_mode
+from torch.distributed._composable.fsdp import FSDP
+from torch.distributed._composable.fsdp._fsdp_param_group import (
+    FSDPCommContext,
+    FSDPParamGroup,
+)
+from torch.distributed._tensor.api import DTensor
+from torch.futures import Future
+from torch.utils._python_dispatch import TorchDispatchMode
+from torch.utils._pytree import tree_map_only
+from torch.utils.hooks import RemovableHandle
+from torch.utils.weak import WeakIdKeyDictionary
+
+_PYTORCH_MIN_ALLOCATE = 2**9
+
+__all__ = ["MemoryTrackingMode"]
+
+
+class _RefType(str, Enum):
+    sharded_parameter = "sharded_parameter"
+    unsharded_parameter = "unsharded_parameter"
+    buffer = "buffer"
+    sharded_gradient = "sharded_gradient"
+    unsharded_gradient = "unsharded_gradient"
+    activation = "activation"
+    all_gather = "all_gather"
+    all_gather_copy_in = "all_gather_copy_in"
+    reduce_scatter = "reduce_scatter"
+    optstate = "optstate"
+    inputs = "inputs"
+
+
+@dataclass
+class _WeakRefInfo:
+    def __init__(
+        self, size: int, element_size: int, reftype: _RefType
+    ) -> None:
+        self.size = size
+        self.element_size = element_size
+        self.reftype = reftype
+        self.mem_consumed = self._calculate_mem_consumed()
+
+    def _calculate_mem_consumed(self) -> int:
+        return (
+            math.ceil((self.size * self.element_size) / _PYTORCH_MIN_ALLOCATE)
+            * _PYTORCH_MIN_ALLOCATE
+        )
+
+    def get_mem_consumed(self, st: torch.UntypedStorage) -> int:
+        if st.size() != self.size:
+            self.size = st.size()
+            self.mem_consumed = self._calculate_mem_consumed()
+        return self.mem_consumed
+
+
+class _FSDPParamGroupSavedMethods(NamedTuple):
+    pre_forward: Callable
+    post_forward: Callable
+    pre_backward: Callable
+    post_backward: Callable
```

Comment:
These methods should be pretty stable, so I am good with patching/introspecting these.